apps/server/src/modules/tldraw/service/tldraw.ws.service.spec.ts

```diff
@@ -0,0 +1,386 @@
+
+	beforeAll(async () => {
+		const imports = [CoreModule, ConfigModule.forRoot(createConfigModuleOptions(config))];
+		const testingModule = await Test.createTestingModule({
+			imports,
+			providers: [TldrawWs, TldrawBoardRepo, TldrawWsService],
+		}).compile();
+
+		service = testingModule.get<TldrawWsService>(TldrawWsService);
+		app = testingModule.createNestApplication();
+		app.useWebSocketAdapter(new WsAdapter(app));
+		jest.useFakeTimers({ advanceTimers: true, doNotFake: ['setInterval', 'clearInterval', 'setTimeout'] });
+		await app.init();
+	});
+
+	afterAll(async () => {
+		await app.close();
+	});
+
+	const createMessage = (values: number[]) => {
+		const encoder = encoding.createEncoder();
+		values.forEach((val) => {
+			encoding.writeVarUint(encoder, val);
+		});
+		encoding.writeVarUint(encoder, 0);
+		encoding.writeVarUint(encoder, 1);
+		const msg = encoding.toUint8Array(encoder);
+		return {
+			msg,
+		};
+	};
+
+	const setupWs = async (docName?: string) => {
+		if (docName) {
+			ws = new WebSocket(`${wsUrl}/${docName}`);
+		} else {
+			ws = new WebSocket(`${wsUrl}`);
+		}
+		await new Promise((resolve) => {
+			ws.on('open', resolve);
+		});
+	};
+
+	it('should service properties be defined', () => {
+		expect(service).toBeDefined();
+		expect(service.pingTimeout).toBeDefined();
+		expect(service.persistence).toBeDefined();
+	});
+
+	describe('when client is not connected to WS', () => {
+		const setup = async () => {
+			await setupWs();
+
+			const closeConSpy = jest.spyOn(service, 'closeConn').mockImplementationOnce(() => {});
+			const sendSpy = jest.spyOn(service, 'send');
+			const doc: { conns: Map<WebSocket, Set<number>> } = { conns: new Map() };
+			const byteArray = new TextEncoder().encode(testMessage);
+
+			return {
+				closeConSpy,
+				sendSpy,
+				doc,
+				byteArray,
+			};
+		};
+
+		it('should throw error for send message', async () => {
+			const { closeConSpy, sendSpy, doc, byteArray } = await setup();
+
+			service.send(doc as WsSharedDocDo, ws, byteArray);
+
+			expect(sendSpy).toThrow();
+			expect(sendSpy).toHaveBeenCalledWith(doc, ws, byteArray);
+			expect(closeConSpy).toHaveBeenCalled();
+
+			ws.close();
+			sendSpy.mockRestore();
+		});
+	});
+
+	describe('when websocket has ready state different than 0 or 1', () => {
+		const setup = () => {
+			const closeConSpy = jest.spyOn(service, 'closeConn');
+			const sendSpy = jest.spyOn(service, 'send');
+			const doc: { conns: Map<WebSocket, Set<number>> } = { conns: new Map() };
+			const socketMock = { readyState: 3, close: () => {} };
+			const byteArray = new TextEncoder().encode(testMessage);
+
+			return {
+				closeConSpy,
+				sendSpy,
+				doc,
+				socketMock,
+				byteArray,
+			};
+		};
+
+		it('should close connection if websocket has ready state different than 0 or 1', () => {
+			const { closeConSpy, sendSpy, doc, socketMock, byteArray } = setup();
+
+			service.send(doc as WsSharedDocDo, socketMock as WebSocket, byteArray);
```

Comment:
For me this are to many as cast. Make it sense to create a testFactory for doc creating to get real type save WsSharedDocDo?

Reply:
Im not sure if we need to do that. This casting is used only in test file for simplifying process of acquiring this object and this mocked object here is not subject of testing.
Those two - WsSahredDocDo (extends from Doc from yjs) and WebSocket are very big classes with lots of properties and methods.

Reply:
as an inbetween, maybe create a "factory" which is performing exactly these casts? then its at least extracted away

Reply:
I've created tldraw.ws.factory.ts as suggested